### PR TITLE
Fix seq2 ite protocol bug

### PR DESCRIPTION
**Pull request by leonardt**
Fix for https://github.com/phanrahan/magma/pull/739

### CODE REVIEW

File: magma/protocol_type.py [hunk 1]
```diff
@@ -86,3 +86,10 @@ def wire(self, other):
 
     def __imatmul__(self, other):
         self.wire(other)
+
+
+def get_type(value):
+    T = type(value)
+    if issubclass(T, MagmaProtocol):
+        return T._to_magma_()
```
Comment (cdonovick) [inline magma/protocol_type.py:94]: Do you want to allow?

```Python
x : ProtocolT
assert ProtocolT._to_magma_() is m.Bits[8]
y : m.Bits[8]
c : m.Bit 
c.ite(x, y)
```
 I do not allow that with the `BitVectorProtocol`

Comment (leonardt) [reply magma/protocol_type.py:94]: Hmm, I think that's the current use case.  Basically, the RegisterWrapper for sequential (implements the call logic) is an implementation of protocol, so this register can be used as a return value (output of any generic type), or you could say, return the register output + 1, or some constant, so for these cases to work, we need the RegisterWrapper (and protocol) to type check with the regular operation (addition result, or constant type)

Comment (cdonovick) [reply magma/protocol_type.py:94]: I think we may want to do a design review at some point about this.  This seems like a very unsafe / surprising coercion.

```Python3
x: Protocol
y: Protocol 
c: Bit

if c:
   z = x
else:
   z = y

assert isinstance(z, Bits)
```

Is not a behaviour I would want.

Comment (leonardt) [reply magma/protocol_type.py:94]: Yea I see the problem with that, let me see how we can workaround this.  Perhaps using an alternative pattern to Protocol would avoid this issue, allowing us to make the type check stricter.  The one issue with sub-classing Bits is that we can't do that in general (the type of a Register can be any arbitrary magma type T), so we would want to dynamically generate subclasses of T that implement the __call__ method, might be complicated but could work and avoid having to relax this type check.

Comment (cdonovick) [reply magma/protocol_type.py:94]: Ahh I see.  That is tricky.  

Perhaps protocols could optionally define a field like `_coercible_` or something if they want to be treated like that?

```Python
def get_type(value):
    T = type(value)
    if issubclass(T, MagmaProtocol) and getattr(T, '_coercible_', False) :
        return T._to_magma_()
    return T
```

Comment (leonardt) [reply magma/protocol_type.py:94]: Yea that could work, basically in this case the register wrapper just wants to behave exactly like it's wrapped typed, it just needs to add some special logic, so "casting" it to a T when it's used in other contexts is totally fine (e.g. the output of the ite), in general a protocol type may not want this, so we could have coercible false be the default

**Pull request merged**